### DISCUSSION TITLE
Potential Performance Improvements

### DISCUSSION THREAD

**Discussion by Peter-Jacob** [adesutherland/CREXX#339] [Ideas]
**1. DO-Loops**
```
loop:
          igt   r1,r3,a2 /* compare counter */
          brt   endloop,r1
          concat r4,r4,a1 /* concat string to return */
          inc   r3
          br    loop
endloop:    
```
we could create super instruction (similar to BCT) for:
```
inc   rx 
br    label 
```
```
loop:
          igt   r1,r3,a2 /* compare counter */
          brt   endloop,r1
          concat r4,r4,a1 /* concat string to return */
          bctp   r3,loop
endloop:    
```
This would require a compiler change.

**Comment by Peter-Jacob** [comment]
I tried the new bctp instruction in a loop:
```
n = 1000000  /* number of iterations    */
do i=1 to n
   j=i
end
```
in both ways and I am a bit confused about the results:
1. conventional do loop 3.247.000 microseconds
2. using bctp 39.992 microseconds. 
this would be 80 times faster. I can't believe this my expectation was 10-20%. Where is my thinking wrong?

**Comment by adesutherland** [comment]
Hi Peter

Surprised it is such an improvement ... definitely need to verify :-) but if true we should definitely implement it. Modern CPUs are weird! 

One thing to remember is that the assembler also optimises by doing instruction replacements in the keyhole optimiser.

So in investigating assembler instruction usage etc. please run the disassembler on the rxbin file. Your tools should be able to use the disassembler output unchanged (fingers crossed)

For example you should see some concats changed to appends.

Also the change you suggest to increment and branch in one instruction could be most easily implemented in the keyhole optimiser

Adrian (in the Chablis market)

**Comment by Peter-Jacob** [comment]
Well, I don't trust my results either. Von meinem Huawei-Telefon gesendet-------- Ursprüngliche Nachricht --------Von: Adrian Sutherland ***@***.***>Datum: So., 11. Juni 2023, 11:50An: adesutherland/CREXX ***@***.***>Cc: Peter-Jacob ***@***.***>, Author ***@***.***>Betreff: Re: [adesutherland/CREXX] Potential Performance Improvements (Discussion #339)
Hi Peter
Surprised it is such an improvement ... definitely need to verify :-) but if true we should definitely implement it. Modern CPUs are weird!
One thing to remember is that the assembler also optimises by doing instruction replacements in the keyhole optimiser.
So in investigating assembler instruction usage etc. please run the disassembler on the rxbin file. Your tools should be able to use the disassembler output unchanged (fingers crossed)
For example you should see some concats changed to appends.
Also the change you suggest to increment and branch in one instruction could be most easily implemented in the keyhole optimiser
Adrian (in the Chablis market)

**Reply by rvjansen** [reply]
If I could get my hands on it I could profile it with perf .. could you push it to a new branch? It could be that the processor cache lines are invalidated by something in the combination of instructions because the C compiler does not trust different threads not to touch it - like in volatile

**Comment by Peter-Jacob** [comment]
I have rewritten the example, and the amazing performance boost has gone. Supposedly, the new BCTP loop counter had a much smaller end limit. Return driver's license at age 75, programming license 65. Sheesh!
The new measurement, for `do i=1 to 10000000`
1. current looping method:   2.56 secs
2. with BCTP 2.43 secs
not much, just 5% better.    
Should we add the BCTP instruction?

**Comment by rvjansen** [comment]
5% when automatically optimizing is not bad at all. (When it is too good to be true, it mostly isn’t). So yes, let’s keep it.René.On 11 Jun 2023, at 19:24, Peter-Jacob ***@***.***> wrote:﻿
I have rewritten the example, and the amazing performance boost has gone. Supposedly, the new BCTP loop counter had a much smaller end limit. Return driver's license at age 75, programming license 65. Sheesh!
The new measurement, for do i=1 to 10000000

current looping method:   2.56 secs
with BCTP 2.43 secs
not much, just 5% better.
Should we add the BCTP instruction?

**Comment by Peter-Jacob** [comment]
**2. DO-Loops part II**

The next improvement could be super instruction:
IGTBR label,value1,value2    if value1>value2 then goto label, else continue

if replaces the following sequence in a do-loop:
```
l15dostart:
   igt r0,r3,r4
   brt l15doend,r0
```
new:
```
l15dostart:
   igtbr l15doend,r3,r4
```
The new measurement, for do i=1 to 10000000

current looping method:   2.56 secs
with BCTP 2.43 secs
with IGTBR 1,87 secs

This adds up to a 27% performance increase

I suggest we keep both instructions internal, to be used in the assembler optimiser. I also think we only support simple do x=1 to ... loops, where the increment is +1. This is the mainly used version anyway. Otherwise we'd need a couple more instructions.

**Comment by Peter-Jacob** [comment]
I used the following REXX:
```
options levelb
n = 10000000  /* number of iterations    */
st=0
et=0
assembler mtime st
do i=1 to n
   j=i
end
assembler mtime et
etf=(et-st)/100000
say etf
say i j
```
after RXC you can replace/add the new instructions in the asm file accordingly.

**Comment by adesutherland** [comment]
Peter 

Feel free to add rules to the keyhole optimiser - assembler/rxas-opt.c line about 355 (detailed notes above it in the c file).

So for 
```
l15dostart:
   igt r0,r3,r4
   brt l15doend,r0
```
to 
```
l15dostart:
   igtbr l15doend,r3,r4
```
The [untested] rule is
```
 *          *  flag type  type     inst     op1     op2     op3
 * input    *  {NO_GAP,   OP_CODE, "igt",   'r', 0, 'r', 1, 'r', 2},
 * input    *  {NO_GAP,   OP_CODE, "brt",   'l', 0, 'r', 0,  0,  0,
 * output 1 *             OP_CODE, "igtbr", 'l', 0, 'r', 1, 'r', 2},
 * output 2 *
 * input    *  {END_OF_RULE},
```

It would be amazing for you to see it work end to end ...

Adrian

**Reply by Peter-Jacob** [reply]
Thanks Adrian, will try it. I work currently on f46, is there a compatability issue when we merge it into dev?

**Reply by adesutherland** [reply]
No, I don't think so - and I am / wont touch the assembler just to make sure!

**Reply by adesutherland** [reply]
I think the 2 'l's need to be 'b' - i.e. for branch not label

**Comment by Peter-Jacob** [comment]
I am close, but losing a label, this is my asm
```
main() .locals=5
   load r1,10
   load r3,1
l15dostart:
   igt r0,r3,r1
   brt l15doend,r0
   bctp l15dostart,r3
l15doend:
   ret
```
after assembly and disasselmbly I get: 
```
.globals=0

main()          .locals=5
                load r1,10                                   * 0x000000:00c3 Load op1 with op2
                load r3,1                                    * 0x000003:00c3 Load op1 with op2
lb_6:           igtbr lb_d,r3,r1                             * 0x000006:0074 Int Greater than if (op2>op3) goto op1
                bctp lb_6,r3                                 * 0x00000a:00ec inc op2; goto op1
lb_d:           ret                                          * 0x00000d:00ae Return VOID
*******************************************************************************
```
the start label disappeared. My rules are:
 ```
         /* do loop improvement: replace igt/brt check by igtbr instruction:     igt r0,r3,r4; brt label,r0 => igtbr label,r3,r4  */
            {ANY_GAP,   OP_CODE, "igt",   'r', 0, 'r', 1, 'r', 2},
            {NO_GAP,   OP_CODE, "brt",   'b', 3, 'r', 4,  0,  0,
                                OP_CODE, "igtbr", 'b', 3, 'r', 1, 'r', 2},
            {END_OF_RULE},

``` 
 I thought maybe I need ANYGAP as first rule entry, but no difference. Do you have any idea?

**Reply by Peter-Jacob** [reply]
going down from rxc/rxas, (then rxdas) all labels are kept. So it seems it is specifically related to asm only.

**Comment by Peter-Jacob** [comment]
After comprehending the optimiser rules I added the following lines to the inline documentation:
```
...
 * d) syntax of the mapping rules
 * *          *  flag type   type     inst         op1 v1  op2 v2  op3 v3
 * input    *  {NO_HAZARD, OP_CODE, "instruction", 'r', 0, 'r', 0, 'r', 1,
 * output 1 *              OP_CODE, "instruction", 'r', 0, 'r', 1,  0, 0},
 * output 2 *
 * input    *  {END_OF_RULE},  ...
 *
 *  op1/op2/op3 are the parameter types of the instruction, r: register, l: label, b: branch
 *  v1/v2/v3 is the temporary variable number in which the parameter content is kept, for the optimising statement.
 *     0:    parameter is not kept
 *     1-10: parameter is kept in the specified variable
 *     This number can be used in the optimised template and allows to merge parts of several input templates
 *
...
```
The optimising for simple do-loops using IGTBR and BCTP are added and tested

**Reply by adesutherland** [reply]
Well done! And sorry I only just picked up your messages.

I forgot that 0 was not a valid 'slot' - to be honest I still can't remember this, I had COVID at the time!

Can you post the rules for future reference?

Adrian

**Reply by adesutherland** [reply]
Also I think it also supports i-integer, s-string, f-float

**Reply by adesutherland** [reply]
PS. The number of people that understand rxas keyhole optimiser in the world has now officially doubled!

**Reply by Peter-Jacob** [reply]
> PS. The number of people that understand rxas keyhole optimiser in the world has now officially doubled!

ha, ha, you are too kind, I doubt I understand it, but I have an idea is the better term ;)

**Reply by Peter-Jacob** [reply]
> Well done! And sorry I only just picked up your messages.
> 
> I forgot that 0 was not a valid 'slot' - to be honest I still can't remember this, I had COVID at the time!
> 
> Can you post the rules for future reference?
> 
> Adrian

These are the 2 new rules:
     ```
     /* do loop improvement: replace igt/brt check by igtbr instruction:     igt r0,r3,r4; brt label,r0 => igtbr label,r3,r4  */
              {ANY_GAP,   OP_CODE, "igt",   'r', 0, 'r', 1, 'r', 2},
            {NO_GAP,   OP_CODE, "brt",   'b', 3, 'r', 4,  0,  0,
                                OP_CODE, "igtbr", 'b', 3, 'r', 1, 'r', 2},
            {END_OF_RULE},
            /*  do loop increase ctr+1 and branch to start   inc rx; br dostart */
            {NO_GAP,   OP_CODE, "inc",   'r', 1, 0, 0, 0, 0},
            {NO_GAP,   OP_CODE, "br",   'b', 2, 0, 0,  0,  0,
                                OP_CODE, "bctp", 'b', 2, 'r', 1, 0, 0},
            {END_OF_RULE},
```
they are pushed to GIT in rxas_opt.c f00046 

I still have trouble understanding when to use NO_GAP and ANY_GAP

**Reply by adesutherland** [reply]
Should the first rule be

{ANY_GAP, OP_CODE, "igt", **'r', 4,** 'r', 1, 'r', 2},
{NO_GAP, OP_CODE, "brt", 'b', 3, **'r', 4**, 0, 0,
OP_CODE, "igtbr", 'b', 3, 'r', 1, 'r', 2},

I.e. r 4 not r 0 - because the register holding the boolean needs to be the same?

Also the first ANY_GAP/NO_GAP makes no difference because there is no preceding instruction (I guess by convention we could use ANY_GAP). Afterwards NO_GAP is what is needed when the two instruction MUST be next to each other - but typically NO_HAZARD should be fine. ANY_GAP only makes sense if you have a label and a branch and don't care what in in between. Clear as mud?

**Reply by Peter-Jacob** [reply]
if I understand the rules correctly, I'd say no. Let me explain, the original code:
 ```
igt r0,r3,r1                     <- r0 contains the boolean result of the comparison r3>r1 **1 
brt l15doend,r0             <- branch based on the boolean result in r0  **2
```
**1 for the new command we need r3 and r1, but not r0, as the branch will be controlled in the IGTBR directly.
Therefore the rule should be `{ANY_GAP, OP_CODE, "igt", 'r', 0, 'r', 1, 'r', 2}, ` drop the boolean register, keep 2. register name in item 1, 3. register in item2

**2 is the branch to replace: `{NO_GAP, OP_CODE, "brt", 'b', 3, 'r', 4, 0, 0,`
we keep the label name in item 3, and drop the boolean part (this is actually wrong in my rule, where I save in item4) 

Replacing the instructions by:
`OP_CODE, "igtbr", 'b', 3, 'r', 1, 'r', 2},`    first part is the label to branch to, if counter>endcounter, this was saved in item 3, next parms are counter (item 1) and endcounter (item 2)

I hope that now everything will be as clear as mud ;)

**Reply by adesutherland** [reply]
But wont your rule incorrectly match something like

igt r99,r3,r1                     <- r99 contains the boolean result of the comparison r3>r1 - r99 presumably used somewhere else in the flow
brt l15doend,r77             <- branch based on the boolean result in r77 - presumably set somewhere earlier

in this case 
igtbr l15doend r3,r1 is wrong

So although r0 is not used in the rewrite it is needed in the selection. It is true that the compiler will never generate such code ...

... and also shows that even the fixed optimiser rule is not perfect - after all handcrafted assembler might expect r0 to be set later but I think this is a risk worth taking. E.g.

igt r0,r3,r1                     
brt l15doend,r0  
xxxx r0 -> cunning hand-crafted assembler which is assuming r0 is 0 (the compiler will never do this)

What do you think about that issue? Worth taking and documenting carefully? Or a future issue that will bite us? We definitely need a wiki page explaining each optimisation

Because these rules will be so confusing to people if they have a bug and the bug so so hard to understand and trace back to the rule - we need to very carefully debate and review any rules. So I hope you don't mind me being anal-retentive!

BTW - at the moment the compiler takes a shortcut and uses r0 often in this situation and otherwise never uses r0 (wasting a register) I am planning to fix this shortly and it will use any spare register.

**Reply by Peter-Jacob** [reply]
you are absolutely right, I didn't think of this! Now the question: you suggested:
```
{ANY_GAP, OP_CODE, "igt", 'r', 4, 'r', 1, 'r', 2},
{NO_GAP, OP_CODE, "brt", 'b', 3, 'r', 4, 0, 0,
```
is specifying the 4 in rule 1 and rule 2 more than just saving the register name, is it also used to match 2 instructions in sequence are candidates for optimisation?

**Reply by adesutherland** [reply]
Yes - that is correct. The example is the swap optimisation that detects swap / swap backs

**Reply by Peter-Jacob** [reply]
indeed it is also used for testing if the instructions belong together! I tried it with:
```
   igt r0,r3,r4
   brt l12doend,r5
```
in this situation, the optimiser doesn't take care. Thanks for the advice!
I will make a note in the program comments.

**Reply by adesutherland** [reply]
Just to explain why the keyhole assembler is so important outside of the compiler. You can imagine that the "igt" is in general just the evaluation of an arbitrary expression. And the 'brt' is part of the loop logic. So for the compiler to emit igtbr it needs an if statement to work out if the expression ended in a "igt" adding to the spaghetti!

However in this case it is not so bad as the loop does indeed know if it is using "igt". This is the loop logic below, the complexity is that it needs to handle incremental loops going up and down - and when it it not know at compile time. e,g,

do a = x to y by 1 --- Up
do a = x to y by -1 --- Down
do a = x to y by z --- Only known at run time when we have evaluated z

The loop handles int or float - e.g. when you see "%sgt" below - this could be igt or fgt. 

(This is also where I need to fix the "r0" hack which means that r0 cannot be used elsewhere by the compiler)

Anyway I can easily change this to emit "igtbr" (the most common loop) and "iltbr", "fgtbr", and "fltbr". If you add these I can edd them  after we merge. What do you think?

```
               /* If the REPEAT has a TO it has an ASSIGN and its register
                 * number will have been set to the ASSIGN Variable */
                node->loopstartchecks = output_fs(comment_meta);
                switch (j) {
                    case 1: /* Positive */
                        temp1 = mprintf("   %sgt r0,%c%d,%c%d\n   brt l%ddoend,r0\n", /* r0 - todo */
                                        tp_prefix,
                                        node->parent->register_type,
                                        node->parent->register_num,
                                        node->child->register_type,
                                        node->child->register_num,
                                        node->parent->parent->node_number);
                        break;
                    case -1: /* Negative */
                        temp1 = mprintf("   %slt r0,%c%d,%c%d\n   brt l%ddoend,r0\n", /* r0 - todo */
                                        tp_prefix,
                                        node->parent->register_type,
                                        node->parent->register_num,
                                        node->child->register_type,
                                        node->child->register_num,
                                        node->parent->parent->node_number);
                        break;
                    default: /* Dynamic by value */
                        /* We need a zero (int or float */
                        if (*tp_prefix == 'i') op = "0";
                        else op = "0.0";
                        temp1 = mprintf(
                                "   %slt r0,%c%d,%s\n" /* Check the by value sign */
                                "   brt l%ddoneg1,r0\n"    /* JMP to Negative BY (r0 - todo) */

                                "   %sgt r0,%c%d,%c%d\n"   /* Pos BY */
                                "   brtf l%ddoend,l%ddoneg2,r0\n" /* r0 - todo */

                                "l%ddoneg1:\n"
                                "   %slt r0,%c%d,%c%d\n"   /* Neg BY */
                                "   brt l%ddoend,r0\n" /* r0 - todo */

                                "l%ddoneg2:\n",
```

**Reply by Peter-Jacob** [reply]
I have no problems adding the new instructions. Give me the marching order!

**Reply by Peter-Jacob** [reply]
I have added "iltbr", "fgtbr", and "fltbr", "igtbr" is already there. 
I actually don't understand what you mean with:
```
/* If the REPEAT has a TO it has an ASSIGN and its register
                 * number will have been set to the ASSIGN Variable */
                node->loopstartchecks = output_fs(comment_meta);
                switch (j) {
                    case 1: /* Positive */
                        temp1 = mprintf("   %sgt r0,%c%d,%c%d\n   brt l%ddoend,r0\n", /* r0 - todo */
                                        tp_prefix,
                                        node->parent->register_type,
                                        node->parent->register_num,
                                        node->child->register_type,
                                        node->child->register_num,
                                        node->parent->parent->node_number);
... 
```
Don't we need the rules then?

**Reply by Peter-Jacob** [reply]
oops, this might be difficult to optimise:
`  if f1>f2 then say "OK"`
becomes:
  ```
 fgt r5,R2,r5
 brf l38iffalse,r5
```
should this become
` flebr l38iffalse,r2,r5 `    not yet installed
instead of a fgtbr?

**Reply by adesutherland** [reply]
I don't think so - wouldn't you keep with gt? Or are you trying to minimize instructions? 

BTW I think the reuse of registers may mean you will need several rules! But what I was saying before (and the code was there to give you a sense of what the compiler should be able to do quite easily) is that I can get the compiler to emit the right igtbr etc. for loops at least

**Reply by Peter-Jacob** [reply]
Let's discuss this in the next biweekly.

**Comment by rvjansen** [comment]
That, by itself, is great news!René.On 13 Jun 2023, at 16:12, Adrian Sutherland ***@***.***> wrote:﻿
PS. The number of people that understand rxas keyhole optimiser in the world has now officially doubled!
